sale_coupon_validity_notification/models/coupon.py

```diff
@@ -0,0 +1,36 @@
+# Copyright 2023 Tecnativa - Pilar Vargas
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import _, models
+
+
+class CouponCoupon(models.Model):
+    _inherit = "coupon.coupon"
+
+    def write(self, vals):
+        res = super().write(vals)
+        if vals.get("state", False) == "expired":
+            for coupon in self.filtered(
+                lambda r: r.state == "expired" and r.sales_order_id
+            ):
+                order = coupon.sales_order_id
+                # Post a note on the chatter and create an activity.
+                order.message_post(
+                    body=_(
+                        "Coupon <a href=# data-oe-model=coupon.coupon data-oe-id=%(id)s>"
+                        "%(coupon)s</a> has expired.",
+                        id=coupon.program_id,
+                        coupon=coupon.program_id.name,
+                    )
+                )
+                order.activity_schedule(
+                    "mail.mail_activity_data_warning",
+                    summary=_("Coupon expired"),
+                    user_id=order.user_id.id or order.partner_id.user_id.id,
```

Comment:
Do you think is a good idea putting an activity to the order's partner? :sweat_smile:

Reply:
Of course not! :woman_facepalming: 
corrected :smiley: